Rust: Implement basic type inference in QL

hvitved opened the pull request:
## Overview

This PR adds basic type inference for Rust, implemented in QL. Type inference and method call resolution are mutually recursive:

```rust
struct<T> MyStruct(T);

impl<S> MyStruct<S> {
    fn foo(self) -> S {
        match self {
            MyStruct(t) => t
        }
    }
}

let x = MyStruct(0);
let y = x.foo();
```

In the example above, in order to resolve the call `x.foo()` we need to know the type of `x`, and in order to resolve the type of `x.foo()` we need to know the return type of `foo` (meaning we need to be able to resolve the call).

The example above involves _generics_, and we want to be able to conclude that `x.foo()` has type `i64` and not simply the type `S`, which is the declared return type of `foo`.

This means that we need to be able to represent the fact that `x` has type `MyStruct<i64>`, and while it is tempting to represent constructed types like this using a `newtype` encoding, such an encoding will become mutually recursive with type inference/call resolution, meaning (at best) poor performance and (at worst) non-monotonic recursion. Another issue with using a `newtype` encoding is that we could only construct a constructed generic type once we are able to infer _all_ type arguments (type arguments will be encoded as cons-lists), and in case one of the type arguments cannot be inferred, we will have to give up completely.

Instead, we index all type resolution relations with a _type path_, which is conceptually a (possibly empty) list of type argument indices. For example, the type `MyStruct<i64>` can be represented as

Type path | Type
----- | ----
`""` | ``MyStruct``
`"0"` | `i64`

Type paths are represented as strings, so there will be no need for additional mutual recursion.

Using type path indexing, we can infer the following types for the program above:

Entity | Type path | Unbound type
------ | ---------- | ------
`self`    | `""`       | `MyStruct`
`self`    | `"0"`      | `S`
`x`    | `""`       | `MyStruct`
`x`    | `"0"`      | `i64`

which means we will be able to match `S` against the type `i64`. We can then combine this with the return type of `foo` to get the correct return type of the call.

## Implementation

The implementation is split into a shared language-agnostic type inference library (in a new `typeinference` QL pack) and a Rust-specific implementation on top.

The shared library defines the `TypePath` class, and provides the module `Matching` for matching types of arguments with types of parameters in the declarations that they target (like matching `i64` with `S` in the example above). Matching takes into account that type arguments may be supplied explicitly (like `MyStruct<i64>::foo(x)`) and that matching may need to take _base types_ into account in order to match. 

The entry predicate of the Rust-specific implementation is
```ql
Type resolveType(AstNode n, TypePath path)
```
which assigns type-path-indexed types to AST nodes. Using the `Matching` module from the shared library, the implementation infers types for record expressions `Foo { bar = baz}`, call expressions `x.foo()` and `Foo::bar(x)`, and field expressions `x.field` (in mutual recursion with `resolveType` in the latter two cases). For call expressions, we take implicit borrows and implicit dereferences into account, and base types in the context of Rust translates to trait bounds and `impl` blocks.

## Known limitations

The implementation has a lot of known limitations, for example:
- Missing support for a lot of expression kinds.
- Missing support for patterns.
- Missing support for operator calls (should be handled like method calls).
- Missing support for constrained `impl` blocks.
- Missing support for `dyn` trait types.
- Missing support for array types.
- Missing support for tuple types.
- Missing support for associated types (should be handled like type parameters, they are the reason for the two new missing data flow results).
- Type matching does not take variance into account; when matching return types, it should be contravariant and not covariant.
- Missing support for the `Deref` trait and more generally [type coercions](https://doc.rust-lang.org/reference/type-coercions.html).
- Missing support for union types

## Evaluation

[DCA shows](https://github.com/github/codeql-dca-main/issues/27146) that the implementation adds a negligible overhead to the total analysis time (~5 %). Overall, we mostly loose call edges compared to what we get from `rust-analyzer` (except for the `diem` project), but this is really not surprising given the known limitations above, as well as the known limitations of our path resolution implementation.

## Note to the reviewer

Commit-by-commit review is encouraged.

paldepind commented on rust/ql/lib/codeql/rust/elements/internal/TypeInference.qll:
```diff
@@ -0,0 +1,999 @@
+ */
+bindingset[path1]
+bindingset[path2]
+private predicate typeSymmetry(AstNode n1, TypePath path1, AstNode n2, TypePath path2) {
```
What about calling this `typeEquality`? That is more descriptive and matches the wording "the same" in the doc. Also this looks like the equality constraints in unification based type inference.

paldepind commented on rust/ql/lib/codeql/rust/elements/internal/TypeInference.qll:
```diff
@@ -0,0 +1,999 @@
+}
+
+/**
+ * Holds if the type of `n1` at `path1` is the same as the type of `n2` at `path2`.
```
```suggestion
 * Holds if the type of `n1` at `path1` is the same as the type of `n2` at
 * `path2` and type information should propagate in both directions through the
 * type equality.
```

paldepind commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,783 @@
+         * For example, in a method call like `M<int>()`, `int` is an explicit
+         * type argument at position `0`.
+         */
+        Type getExplicitTypeArgument(TypeArgumentPosition tapos, TypePath path);
```
I'm a little unsure about the "explicit" here. It makes sense for C# and Rust that have implicit and explicit type arguments. But there are also languages that have implicit type parameters, and we don't have the word "explicit" in `getTypeParameter`. There's also languages that have implicit normal (non-type) arguments, and we don't call normal arguments "explicit". So maybe going for the more generic name without "explicit" is better for a language agnostic interface?

paldepind reviewed (changes requested):
This is fantastic work and it's very interesting for me to see how one would implement type inference in QL.

My understanding is that at a high-level the type inference works by starting with types that are known (usually from declarations) and then having those type propagate though the AST (forwards and backwards). A big part of that is the `Matching` module which pairs declarations with usages of the declaration, and propagates declared types from the declaration to the use.

This is somewhat different from traditional type inference by unification where one builds bigger and bigger sets of terms that have equal type, with the hope that the set is eventually unified with a fully concrete type. However, in QL we can never "update" anything in place, and hence the present approach works better as all the sub-steps produce final accurate information, coupled with the path approach that makes it possible to build different layers of a type in smaller steps.

Overall the PR looks great and I've only "stylistic" comments. There are still things that I'd like to look closer at, but IMO it makes the most sense that I just keep doing that and create follow up PRs for any further suggestions I might have.

In addition to the comments, here are two high-level thoughts:

* We already use "resolution" for matching paths to declarations. In this PR "resolve" often seems to be synonymous with "infer". Could we instead use "infer" and reserve "resolve" for path resolution only? So for instance `resolveType` would be `inferType`?

Second thought is more regarding potential follow up work:

* In C# constraints on types are based on classes and interfaces, both of which are types themselves. So a constraint is something like "type A is a subtype of type B" whereas in Rust constraints are "type A implements trait T". That is, it's a not a `type * type` relation but a `type * trait` relation. Currently this is handled by treating traits as types and trait implementation like subtyping.

  I wonder if things could be a bit clearer by instead introducing a `TypeConstraint` class in the shared interface as a generalization. Perhaps this could also reduce the number of types produced in examples like the one in the documentation for `resolveType/2` where a term has both a type and a trait as its type. Unless there's a reason why this wouldn't work, this is something that I'd like to play around with.

hvitved commented:
> My understanding is that at a high-level the type inference works by starting with types that are known (usually from declarations) and then having those type propagate though the AST (forwards and backwards). A big part of that is the `Matching` module which pairs declarations with usages of the declaration, and propagates declared types from the declaration to the use.
> 
> This is somewhat different from traditional type inference by unification where one builds bigger and bigger sets of terms that have equal type, with the hope that the set is eventually unified with a fully concrete type. However, in QL we can never "update" anything in place, and hence the present approach works better as all the sub-steps produce final accurate information, coupled with the path approach that makes it possible to build different layers of a type in smaller steps.

Your understanding is absolutely correct.
 
> * We already use "resolution" for matching paths to declarations. In this PR "resolve" often seems to be synonymous with "infer". Could we instead use "infer" and reserve "resolve" for path resolution only? So for instance `resolveType` would be `inferType`?

Good idea, I'll change it.

> * In C# constraints on types are based on classes and interfaces, both of which are types themselves. So a constraint is something like "type A is a subtype of type B" whereas in Rust constraints are "type A implements trait T". That is, it's a not a `type * type` relation but a `type * trait` relation. Currently this is handled by treating traits as types and trait implementation like subtyping.
>   I wonder if things could be a bit clearer by instead introducing a `TypeConstraint` class in the shared interface as a generalization. Perhaps this could also reduce the number of types produced in examples like the one in the documentation for `resolveType/2` where a term has both a type and a trait as its type. Unless there's a reason why this wouldn't work, this is something that I'd like to play around with.

I don't know if this could work, but feel free to try it out. Note that we also need type constraints for traits that extend other traits.

hvitved commented on rust/ql/lib/codeql/rust/elements/internal/CallExprBaseImpl.qll:
```diff
@@ -30,14 +26,7 @@ module Impl {
-      result = resolvePath(this.(CallExpr).getFunction().(PathExpr).getPath())
-    }
+    /** Gets the static target of this call, if any. */
+    Callable getStaticTarget() { none() }
```
No, it is because the predicate is defined by virtual dispatch, I'll add a comment stating so.

paldepind replied:
What I mean is, the `none()` implementation is never used, since both `CallExpr` and `MethodCallExpr` implement `getStaticTarget` themselves. So ideally we'd just have
```diff
-   Callable getStaticTarget() { none() }
+   abstract Callable getStaticTarget();
```
but `CallExprBase` is extended by the generated `CallExpr` (where we can't add `getStaticTarget`) and not the hand-modifiable wrapper `CallExpr` (where we have added `getStaticTarget`) and therefore we're forced to add a stub default implementation.

So the `none()` is an artifact of how we generate classes, which I thought was worth mentioning in a comment :)

hvitved replied:
Yes, now I understand, it is correct that we cannot make it `abstract`.

hvitved commented on rust/ql/lib/codeql/rust/elements/internal/PathResolution.qll:
```diff
@@ -276,18 +304,73 @@ class ImplItemNode extends ImplOrTraitItemNode instanceof Impl {
 
+  pragma[nomagic]
+  private TypeParamItemNode getASelfTyTypeParamArg(TypeRepr arg) {
+    arg = this.getASelfTyArg() and
```
Not sure I understand, if we remove this line it will introduce a Cartesian product with `this`.

paldepind replied:
Yeah, sorry, I was not very clear. Since the first conjunct is always true where the predicate is used, it can be removed. And since that's is the only place where `this` is used, the predicate could be a non-member predicate instead:
```ql
private TypeParamItemNode resolvePathTypeRepr(TypeRepr arg) {
  result = resolvePath(arg.(PathTypeRepr).getPath())
}
```
I think this predicate could also be used inside `hasTraitBound` [here](https://github.com/hvitved/codeql/blob/rust/type-inference/rust/ql/lib/codeql/rust/internal/PathResolution.qll#L503):
```diff
-       this = resolvePath(wp.getTypeRepr().(PathTypeRepr).getPath()) and
+       this = resolvePathTypeRepr(wp.getTypeRepr()) and

hvitved replied:
Ah, yes, makes sense.

hvitved commented on rust/ql/lib/codeql/rust/elements/internal/PathResolution.qll:
````diff
@@ -385,12 +480,50 @@ private class TypeParamItemNode extends ItemNode instanceof TypeParam {
+   * ```
+   */
+  pragma[nomagic]
+  predicate isConstrained() {
````
How about `isFullyParametric`?

paldepind replied:
Those are great 👍

hvitved commented on rust/ql/lib/codeql/rust/elements/internal/StructImpl.qll:
```diff
@@ -32,5 +32,17 @@ module Impl {
+    /**
+     * Holds if this struct uses record fields.
+     *
+     * Empty structs are considered to use record fields.
```
I'll update the tests to illustrate the difference.

hvitved commented on rust/ql/lib/codeql/rust/elements/internal/TypeInference.qll:
```diff
@@ -0,0 +1,999 @@
+private Type resolvePathExprType(PathExpr pe, TypePath path) {
+  // nullary struct/variant constructors
+  not exists(CallExpr ce | pe = ce.getFunction()) and
+  path.isEmpty() and
```
That should hopefully be handled by the context in which the `None` occurs.

paldepind replied:
I see. Then I should probably just play around with this to see how it works.

hvitved commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,783 @@
+
+    /** Holds if this path starts with `tp`, followed by `suffix`. */
+    bindingset[this]
+    predicate startsWith(TypeParameter tp, TypePath suffix) {
```
I'll go for `isCons`. We actually don't use `endsWith`, so I'll remove it.

hvitved commented on rust/ql/lib/codeql/rust/elements/internal/PathResolution.qll:
```diff
@@ -83,6 +83,9 @@ abstract class ItemNode extends AstNode {
   abstract Visibility getVisibility();
 
+  /** Gets the `i`th type parameter of this item, if any. */
+  abstract TypeParam getTypeParam(int i);
```
I think it is better to be explicit.

hvitved commented:
@paldepind : Thanks for the review, I have (hopefully) addressed all your comments. I also decided to rebase, so we can see the combined effects with https://github.com/github/codeql/pull/18228 on DCA.

paldepind reviewed (approved):
Thanks for addressing my comments :) Looks great to me 🎉 🎉

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+       */
+      bindingset[a, target]
+      pragma[inline_late]
+      private Type getTypeArgument(Access a, Declaration target, TypeParameter tp, TypePath path) {
```
Would it make sense to add the context `a.getTarget() = target` to this predicate, and drop the inlining? It seems to depend on that context in all use-cases, and we'll materialize that combination anyway in `explicitTypeMatch`. That way the two negated occurrences would become simple anti-joins, and we could drop `explicitTypeMatch` (or rename this predicate to `explicitTypeMatch`).

paldepind replied:
Makes sense to me.

hvitved replied:
This will add more recursion, since `getTarget` is often mutually recursive with type inference, so that's why I did it like this. But it could deserve a comment saying so.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+         * `baseMention` | `path`    | `t`
+         * ------------- | --------- | ---
+         * `Mid<C<T4>>`  | `"0"`     | ``C`1``
+         * `Mid<C<T4>>`  | `"0.1"`   | `int`
+         * `Base<C<T3>>` | `"0"`     | ``C`1``
+         * `Base<C<T3>>` | `"0.0"`   | ``C`1``
+         * `Base<C<T3>>` | `"0.0.1"` | `int`
```
Are these `1`s correct? Given that the examples so far have used positional, zero-based indices, shouldn't these `1`s be `0`s instead?

hvitved replied:
Yes 🤦

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
````diff
@@ -0,0 +1,776 @@
+         * new Sub<int>().ToString();
+         * ```
+         *
+         * for the node `new Sub<int>()`, which is the receiver of a method call, we
+         * have:
````
Could we clarify this a bit? It's somewhat confusing which `Access` that the example concerns: If I've understood the concepts correctly, there are two `Access`es, one that's a constructor call and one that's a call to a method named `ToString`. The former has maybe a `TypeArgument` with a `TypeArgumentPosition`, or is it simply a `TypeMention` with `int` occurring at `TypePath` `"0"`. Or is it both? The second `Access` has a single argument, which is also the qualifier, and that argument/qualifier plays a crucial role in determining the call target. So if call target resolution (that's an input to this module) is mutually recursive with this output, then it looks odd that this predicate has `relevantAccess` and hence a call target as a dependency, since that smells like something that might end up as en empty recursion?

hvitved replied:
Yes, it concerns the second access.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+        /**
+         * Gets the declared type of this declaration at `path` for position `dpos`.
+         *
+         * For example, if this declaration is the method `int M(bool b)`,
+         * then the declared type at parameter position `0` is `bool` and the
+         * declared return type is `int`.
```
Could we elaborate the example concerning the following? If `M` is an instance method then presumably there's also a `DeclarationPosition` for the instance argument? If not, then the `new Sub<int>().ToString();` example below in `hasBaseTypeMention` would not qualify for `predicate relevantAccess`.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+        Declaration decl, DeclarationPosition dpos, Type base, TypePath path, Type t
+      ) {
+        t = decl.getDeclaredType(dpos, path) and
+        path.isCons(base.getATypeParameter(), _)
```
This relies crucially on `TypeParameter`s being unique to a given `Type`. Should that be checked in a consistency check somewhere?

hvitved replied:
The rewrite will not work if a declaration has multiple declared types at `dpos`, which there is no reason to rule out. But I noticed that `declarationBaseType` is only used in a context where `t` is in fact a `TypeParameter`, so we may as well specialize it.

hvitved replied:
> This relies crucially on TypeParameters being unique to a given Type. Should that be checked in a consistency check somewhere?

Good idea; not all type parameters need to belong to a type though (e.g. method type parameters in C#), but they should belong to at most one type.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+   * `"1.0"`    | `C`
+   * `"1.1"`    | `D`
+   *
+   * Note that while we write type paths using type parameter positions (e.g. `"0.1"`),
+   * the actual implementation uses unique type parameter identifiers, in order to not
+   * mix up type parameters from different types.
```
Or we could even go so far as saying that for the purposes of the examples we assume that `T_i` maps to the integer `i`, and then we'd get e.g. `"1.4"` pairs with `B`. Then we'd match the actual implementation even more.

paldepind replied:
I think that is a good idea. I think the first would be easiest to read, and it does match how type paths are `toString`'ed as well.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        exists(AccessPosition apos, DeclarationPosition dpos, Type base, TypePath pathToTypeParam |
+          accessBaseType(a, apos, target, base, pathToTypeParam.append(path), t) and
+          declarationBaseType(target, dpos, base, pathToTypeParam, tp) and
```
To check my understanding: The join of `base` here is the point where we introduce covariance, correct? Perhaps leave a comment in this spot about covariance if this is the place that needs updating to potentially support contravariance in the future?

hvitved replied:
Correct.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+      private predicate implicitTypeMatch(
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        directTypeMatch(a, target, path, t, tp)
+        or
+        baseTypeMatch(a, target, path, t, tp)
```
So `adjustAccessType` is applied in `directTypeMatch`, but not in `baseTypeMatch`? Maybe an example to clarify why we have this distinction would be good to add in the qldoc for `adjustAccessType`.

hvitved replied:
Correct, only used in `directTypeMatch`. I think we would have to add it in `accessBaseType` as well.

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+      }
+
+      /**
+       * Holds if the (transitive) base type `t` at `path` of `a` matches the type
```
Maybe update
```
at `path` of `a`
```
to
```
at `path` of some `AccessPosition` of `a`
```

aschackmull commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+       *
+       * `path`      | `t`
+       * ----------- | -------
+       * `"0"`       | ``Base`1``
+       * `"0.0"`     | ``C`1``
+       * `"0.0.0"`   | ``C`1``
+       * `"0.0.0.1"` | `int`
```
same

hvitved commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+       *
+       * `path`    | `t`
+       * --------- | -------
+       * `"0"`     | ``C`1``
+       * `"0.0"`   | ``C`1``
+       * `"0.0.1"` | `int`
```
Yes, good spot.

hvitved commented on shared/typeinference/codeql/typeinference/internal/TypeInference.qll:
```diff
@@ -0,0 +1,776 @@
+       *
+       * `path`      | `t`
+       * ----------- | -------
+       * `"0"`       | ``C`1``
+       * `"0.0"`     | ``C`1``
+       * `"0.0.0"`   | ``C`1``
+       * `"0.0.0.1"` | `int`
```
Yes, I really managed to get those examples wrong...
